rootless: automatically create a systemd scope

giuseppe opened the pull request:
when running in rootless mode and using systemd as cgroup manager create automatically a systemd scope.

This solves a couple of issues:

on cgroup v2 it is necessary that a process before it can moved to a different cgroup tree must be in a directory owned by the unprivileged user.  This is not always true, e.g. when creating a session with su -l.

Closes: https://github.com/containers/libpod/issues/3937

Also, for running systemd in a container it was before necessary to specify "systemd-run --scope --user podman ...", now this is done automatically as part of this PR.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

mheon commented on libpod/oci_internal_linux.go:
```diff
@@ -367,7 +367,7 @@ func (r *OCIRuntime) moveConmonToCgroupAndSignal(ctr *Container, cmd *exec.Cmd,
 		logrus.Infof("Running conmon under slice %s and unitName %s", realCgroupParent, unitName)
 		if err := utils.RunUnderSystemdScope(cmd.Process.Pid, realCgroupParent, unitName); err != nil {
-			logrus.Warnf("Failed to add conmon to systemd sandbox cgroup: %v", err)
+			logrus.Warningf("Failed to add conmon to systemd sandbox cgroup: %v", err)
```
?

giuseppe replied:
Thanks for spotting it :) got confused.  Dropped the patch

mheon commented on cmd/podman/main_local.go:
```diff
@@ -139,6 +142,32 @@ func setupRootless(cmd *cobra.Command, args []string) error {
+			return err
+		}
+		unitName := fmt.Sprintf("podman-%d.scope", os.Getpid())
+		if err := utils.RunUnderSystemdScope(os.Getpid(), "user.slice", unitName); err != nil {
```
Should this be unconditional? I feel like we only want it with systemd CGroups.

giuseppe replied:
we also want this when using cgroupfs.  It solves the problem of running systemd containers on cgroups v1 with: systemd-run --scope --user podman ...

If it fails though, we just print a debug message

rhatdan commented:
LGTM

rhatdan commented:
@msekletar  PTAL

TomSweeneyRedHat commented:
LGTM assuming happy tests

giuseppe commented:
tests are finally passing

rhatdan commented on pkg/cgroups/cgroups_supported.go:
```diff
@@ -25,3 +32,58 @@ func IsCgroup2UnifiedMode() (bool, error) {
+		if err != nil {
+			return false, err
+		}
+		s := st.Sys()
```
Nit: why are you creating s?

rhatdan commented on cmd/podman/main_local.go:
```diff
@@ -119,7 +123,31 @@ func profileOff(cmd *cobra.Command) error {
+		return errors.Wrapf(err, "cannot parse pid file %s", pausePidPath)
+	}
+
+	unitName := "podman-pause.scope"
```
Nit: Why are you creating unitName?

mheon commented:
I don't like increasing our dependency on systemd; I'd like to still be able to run in environments where it's not available. Still, this doesn't seem harmful... LGTM

giuseppe commented:
I've addressed the comments and pushed a new version.



> I don't like increasing our dependency on systemd; I'd like to still be able to run in environments where it's not available. Still, this doesn't seem harmful... LGTM

I agree, Podman should work fine without systemd.  If it fails to create the scope we only give a warning (if the cgroups manager is `systemd`) or a debug message (with `cgroupfs`).

rhatdan commented:
@giuseppe needs a rebase.
